bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py

```diff
@@ -77,20 +73,72 @@ def RDataFrameAsNumpy(df, columns=None, exclude=None):
         column_type = df.GetColumnType(column)
         result_ptrs[column] = df.Take[column_type](column)
 
-    # Convert the C++ vectors to numpy arrays
-    py_arrays = {}
-    for column in columns:
-        cpp_reference = result_ptrs[column].GetValue()
-        if hasattr(cpp_reference, "__array_interface__"):
-            tmp = numpy.asarray(cpp_reference) # This adopts the memory of the C++ object.
-            py_arrays[column] = ndarray(tmp, result_ptrs[column])
-        else:
-            tmp = numpy.empty(len(cpp_reference), dtype=numpy.object)
-            for i, x in enumerate(cpp_reference):
-                tmp[i] = x # This creates only the wrapping of the objects and does not copy.
-            py_arrays[column] = ndarray(tmp, result_ptrs[column])
+    result = AsNumpyResult(result_ptrs, columns)
+
+    if lazy:
+        return result
+    else:
+        return result.GetValue()
 
-    return py_arrays
+
+class AsNumpyResult(object):
+    """Future-like class that represents the result of an AsNumpy call.
+
+    Provides AsNumpy with laziness when it comes to triggering the event loop.
+
+    Attributes:
+        columns (list): list of the names of the columns returned by
+            AsNumpy.
+        py_arrays (dict): results of the AsNumpy action. The key is the
+            column name, the value is the NumPy array for that column.
+        result_ptrs (dict): results of the AsNumpy action. The key is the
+            column name, the value is the result pointer for that column.
+    """
+    def __init__(self, result_ptrs, columns):
+        """Constructs an AsNumpyResult object.
+
+        Parameters:
+            result_ptrs (dict): results of the AsNumpy action. The key is the
+                column name, the value is the result pointer for that column.
+            columns (list): list of the names of the columns returned by
+                AsNumpy.
+        """
+
+        self.result_ptrs = result_ptrs
+        self.columns = columns
+        self.py_arrays = None
```

Comment:
I asked myself what to do with attributes in this class and also in `CppWorkflow` w.r.t. naming conventions. I find it a bit annoying to have to prepend every data attribute with `_` if you only want to access them via methods. In `CppWorkflow` I decided to only flag as public/private the methods and leave alone the data attributes. I can add the `_` flag here since it's a class that will be exposed to the user - to be discussed what we do with RDF's internal classes.

Reply:
>  I find it a bit annoying to have to prepend every data attribute with `_`

if they are internal, that's what PEP 8 prescribes.

again alternatively we can just make it so that it's ok to access the attributes so we can keep them public.

Reply:
No I think it's better that they just stay private - I'll flag them as such.